src/main/scala/viper/gobra/translator/encodings/combinators/TypeEncoding.scala

```diff
@@ -403,3 +433,52 @@ trait TypeEncoding extends Generator {
   protected def synthesized[T](node: (vpr.Position, vpr.Info, vpr.ErrorTrafo) => T)(comment: String): T =
     node(vpr.NoPosition, vpr.SimpleInfo(Seq(comment)), vpr.NoTrafos)
 }
+
+object TypeEncoding {
+
+  import viper.gobra.translator.util.ViperWriter.{CodeLevel => cl}
+
+  /**
+    * Checks whether an L-value is safe, i.e. does not cause a runtime panic due to dereferencing nil.
+    *
+    * assert [&loc != nil: *T°]; res
+    *
+    */
+  final def checkNotNil(loc: in.Location, res: vpr.Exp)(ctx: Context): CodeWriter[vpr.Exp] = {
+    if (cannotBeNil(loc)) cl.unit(res)
+    else {
+      for {
+        cond <- checkNotNil(loc)(ctx)
+        checked <- cl.assertWithDefaultReason(cond, res, DerefError)(ctx)
+      } yield checked
+    }
+  }
+
+  /**
+    * Checks whether an L-value is safe, i.e. does not cause a runtime panic due to dereferencing nil.
+    *
+    * [&loc != nil: *T°]
+    *
+    */
+  final def checkNotNil(loc: in.Location)(ctx: Context): CodeWriter[vpr.Exp] = {
+
+    val annotatedInfo = loc.info.asInstanceOf[Source.Parser.Single].createAnnotatedInfo(Source.ReceiverNotNilCheckAnnotation)
+    val (pos, info, errT) = loc.vprMeta
+
+    if (cannotBeNil(loc)) cl.unit(vpr.TrueLit()(pos, info, errT))
+    else {
+      ctx.expression(in.UneqCmp(
+        in.UncheckedRef(loc)(annotatedInfo),
+        in.NilLit(in.PointerT(loc.typ, Exclusive))(annotatedInfo)
+      )(annotatedInfo))
+    }
+  }
+
+  @tailrec
+  private def cannotBeNil(l: in.Expr): Boolean = l match {
```

Comment:
Is this meant to only check locations? If so, I would probably make the param type more restrictive (i.e., `in.Location`)

Reply:
No, it also checks expressions (in particular, see the body of the method, where we use it to check whether the receiver of an index and field-ref expression cannot be nil).

Reply:
I could change it to something along the lines of:
```
private def cannotBeNil(l: in.Location): Boolean = {
  def aux(e: in.Expr): Boolean = {
    // body of cannotBeNil
  }
  aux(l)
}
```